DateMath: Interactivity

GuiltyDolphin opened the pull request:
This adds interactivity to the DateMath Goodie through JavaScript.

_Hide your eyes._
### To-Do
- [x] Use a single line for the amount input, in the form `Years: X Months: X etc.` with support for changing the operation from addition to subtraction.

---

IA Page: https://duck.co/ia/view/date_math

daxtheduck commented:
Metadata placeholder post
            
*If you haven't already, please edit your first comment to include a link to the IA page at the bottom.*
            
---
            
**How to find your IA page**
- New Instant Answer
  -- follow the steps [here](https://duck.co/ia/new_ia) to create your IA page.
              
- Existing Instant Answer
  -- find your IA page from this [list](https://duck.co/ia)

GuiltyDolphin commented:
@mintsoft Want to have a look over this? I think the functionality is all there; just a question of getting it to look nice.

I'll have another look pretty soon to make sure I've removed everything that should be, and clean up anything that requires it.

mintsoft commented:
@GuiltyDolphin certainly can

mintsoft commented:
Hmm so I'm a smidge confused by the GUI at first; I've just played with it and it made more sense once I could see how it reacted.

@abeyang Any design thoughts here?

GuiltyDolphin commented:
@mintsoft Glad it made sense (eventually)! I'm no designer so any design review is appreciated.

mintsoft commented:
@GuiltyDolphin Immediate things that I've noticed are that I think some labels might help?
* It'd be nice to highlight the "start date"
* The input boxes on the start bit are a little difficult to tell that you can edit them, because some bits you can, some you can't
* Personally I dislike month first date formats; I'd prefer like `Monday 25th April 2016 21:17:05`

mintsoft commented:
Also it doesn't give me the answer when I query? I have to tinker with it to get the answer out?

GuiltyDolphin commented:
Oh lord @mintsoft yeah - forgot to turn that on!

GuiltyDolphin commented:
@mintsoft Sure - I was using a random format suggested by `moment.js`. Probably best to go with the general 'DuckDuckGo' format (with weekday).

mintsoft commented:
Ah somehow I knew it'd be `moment.js`'s fault

GuiltyDolphin commented:
@mintsoft I've updated the date format and fixed the date not being initially calculated.

For the 'indication' that the fields are editable; I'm pretty happy with the current 'neutral' style for the date, so how about a [popout](https://duckduckgo.com/styleguide#popouts) or tooltip?

mintsoft commented:
@GuiltyDolphin I think my confusion around the input boxes is because that line is small and you highlight things like the sunday it makes it look like it's just text. Maybe if you did ` user-select: none; ` (and all the prefixes for different browsers) it'd be more obvious? I dunno

mintsoft commented:
Hmm maybe change them to `<input type='number' />` ? Then they'll have little spinners on them too; makes it a bit clearer when you click it

GuiltyDolphin commented:
@mintsoft I had that (spinners) before - it was pretty ugly.

There is the indication when you hover over (a caret); but I guess we could make the highlighting a bit more prominent - I'll see how it looks.

moollaza commented:
Just looking over this. My initial reaction is that there's too much JS here. It looks like a lot of what you've implemented can be handled by the browser:

- Changing the input `type` to number will and setting `max=`, `min=` will help a lot
- That should force only acceptable inputs, meaning no more validation required
- Our `btn` class should handle all the UI aspects of any buttons: https://duckduckgo.com/styleguide#buttons
- Our grid classes should help with the layout -- though I suspect our design team will have thoughts on the overall UI

moollaza commented on share/goodie/date_math/date_math.js:
```diff
@@ -0,0 +1,372 @@
+            'Second', 'Minute', 'Hour',
+            'Day', 'Week', 'Month', 'Year'
+        ];
+        function pad(text, padChar, padWidth) {
```
Move private functions outside the `build` function definition

moollaza commented on share/goodie/date_math/date_math.js:
```diff
@@ -0,0 +1,372 @@
+                    $minute = $dom.find('.input--minute'),
+                    $second = $dom.find('.input--second');
+
+                $dom.find('form').submit(function(e) {
```
looks like you can remove this -- we're not using a `<form>`

GuiltyDolphin replied:
Yeah, I think I was before - there may be a few other legacy pieces, let me know if you spot any!

moollaza commented on share/goodie/date_math/date_math.js:
```diff
@@ -0,0 +1,372 @@
+                        var minute = $minute.val();
+                        var second = $second.val();
+                        var time = hour + ':' + minute + ':' + second;
+                        var date = moment(year + "-" + month + "-" + day +
```
I would use `moment().set({'year': 2013, 'month': 3}).format(...);` or `moment().year(NNNN).month(NN).date(NN).format(.....)`

GuiltyDolphin commented:
@moollaza 

https://github.com/duckduckgo/zeroclickinfo-goodies/pull/2897/commits/34ca44f756e4d0514b577194dc43c1111e8d365d - I was using `number` with `min` and `max` before; but I think it was causing some display issues; I'll check it out again.

I am using `btn` - but there's some overriding going on that made it a little painful and I had to add some custom stuff... Some design advice/feedback would be very welcome!

Thanks for the review :smile:

GuiltyDolphin commented:
@moollaza BTW, which (specifically) were you thinking should be switched to `number`? The `day` is supposed to allow inputs like `2nd`, `3rd` etc., the amount ends up looking like:

![screenshot-2016-06-02-21 12 42_598_47](https://cloud.githubusercontent.com/assets/8598426/15759514/d00420e4-2906-11e6-8a94-89eea38a400e.png)

The times get really screwed up as well.

Am I missing something? It feels like there should be a way to *not* have it screw everything up and just have the input type; though looking at other IAs they've used things like `tele` and `text` inputs to bypass this.

GuiltyDolphin commented:
@moollaza Ah, playing around with styles a bit I think I might be able to replace some of the `text` input with `number`.

GuiltyDolphin commented:
@moollaza Right, I've changed some of the fields to use `tel` and `number`, and I've removed a bunch of the custom validity checking; allowing the removal of a fair bit o' JS.

I think I preferred the old checks, but this should serve :+1:

In terms of functionality, I think this is there - I think the design needs touching up a 'little', but apart from that... This should be fairly complete.

Would it be possible to get this on Beta soonish?

moollaza commented:
> Right, I've changed some of the fields to use tel and number, and I've removed a bunch of the custom validity checking; allowing the removal of a fair bit o' JS.

Hmm I'm not a fan of using `tel` for anything that isn't actually a **tel**ephone number -- that's just abusing HTML.

I plan to speak to the team about how we can better handle all these forms and inputs, and also handle DOM changes. There's no shortage of JS validation libs and I don't want us to have to implement our own checks in each IA. Similarly we should be able to bind all these inputs to model so we don't need to keep searching the DOM and modifying the `html()` and `val()` etc. That's just icky.

>The day is supposed to allow inputs like 2nd, 3rd etc., the amount ends up looking like:

Why do we need to handle inputs like "2nd", "3rd", etc?

GuiltyDolphin commented:
@moollaza Uhh... We don't need to handle inputs like `2nd`, `3rd` etc., that was my bad - it just normalizes to that (also meaning that `number` doesn't work).

Whoops - yeah, we can just use 'text' for a bunch of these (as before); just with less checks.

moollaza commented:
Updated on Beta.

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+  </span>
+  <span class="input--op g thirty">
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select gw">
```
remove this `gw`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select gw">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="g five" />
```
remove this `<span>`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+<div class="gw">
+  <span class="input--date">
```
add `sixty`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+<div class="gw">
+  <span class="input--date">
+    <span class="frm g thirty">
```
change to `sixty`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
+    </span>
+    <span class="frm input--time input--numeric g thirty">
```
change to `forty`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+  <span class="input--date">
+    <span class="frm g thirty">
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
```
add a CSS rule for this to have  `width: 100%`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+    <span class="frm op--select gw">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="g five" />
+      <span class="frm__select g fifty">
```
need to add css to set `height: 2em` might need `!important`

GuiltyDolphin replied:
@moollaza Hmm, tried setting the style via CSS (and adding IDs, and a bunch more classes to try and improve chances of being selected) but no luck (with or without `!important`); works if I use `style="height:2em"` though.

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="g five" />
+      <span class="frm__select g fifty">
+        <select class="input--op-type" />
```
same here, need to add css to set `height: 2em`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,38 @@
+<div class="gw">
```
Give this a class like `.container`

add css for `margin-bottom: 1em`

moollaza commented:
@GuiltyDolphin this is looking amazing 👍 💯  Nicely done.

I've left some comments about CSS fixes that should be made. The Pikaday css is slightly broken causing the modal to have no background, but I've already made a simple internal PR. Once that's deployed to beta the pikaday selector will work just fine.

The end result of my CSS looks like this:

![jan_1_2012_plus_32_days_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/16932722/37eaf6a0-4d16-11e6-9f1a-2822407e8fd7.png)

GuiltyDolphin commented:
@moollaza Thanks!

I've updated the design per your suggestions - it looks better now but my version appears to have a couple of issues still (not sure if this is just me). Some of the external CSS *refused* to be applied to the elements so I've had to use `style="..."` on some to *force* the changes.

tagawa commented:
I just installed the updated version to beta: https://beta.duckduckgo.com/?q=Jan%201%202012%20plus%2032%20days

<img width="400" alt="datetime_ia_screenshot" src="https://cloud.githubusercontent.com/assets/94173/17145858/037d8df6-5397-11e6-829e-21907ac43141.png">

Looks like we may need to change the class or style for the Result line, and the length of the calendar field. Would using `type=date` for the input help there?

tagawa commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,37 @@
+      <p class="frm__desc tx--11">TIME</p>
+      <input type="text" maxlength='2' pattern='^(2[0-3]|[0-1]?[0-9])$'
+        size='2' class="input--hour frm__input" />
+      <span class="sep--time">:</span>
```
Would it be easier to use `<input type="number" min=0 max=23 class="input--hour frm__input" />` here, and similar for the subsequent fields?

GuiltyDolphin replied:
@tagawa I've had a few issues with the `number` input type, the most pressing in this case being that you cannot (easily) normalize the format to two digits (e.g., `7` -> `07`); so not really :/

tagawa replied:
@GuiltyDolphin Understood. Please ignore my comment.

tagawa commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,37 @@
+  <span class="input--date sixty">
+    <span class="frm g sixty">
+      <p class="frm__desc tx--11">DATE</p>
+      <input type="text" class="frm__input date--start" />
```
Would `type="date"` be easier here? It would avoid the text truncation issue if so.

GuiltyDolphin replied:
@tagawa I don't get any wrapping locally (though the rest of the display is a bit off):

![screenshot-2016-07-28-10 45 45_876_268](https://cloud.githubusercontent.com/assets/8598426/17208422/95d2416e-54b0-11e6-99b2-2e8d44c1e669.png)

Though switching it to `date` doesn't seem to make a difference, so I'll try that.

GuiltyDolphin replied:
Reverted due to browser issues.

GuiltyDolphin commented:
@tagawa I've fixed the result line color and updated the input field to `date` - with any luck that'll fix it on beta (though I didn't notice any difference locally).

tagawa commented:
Thanks @GuiltyDolphin. The dark theme (and other themes) look good now, and personally I find the date input easy to use. A couple of points:

1. The three Time boxes are wrapping in Firefox for me. Safari and Vivaldi are fine.

2. The output date is not showing in Vivaldi (and I assume Opera, Chrome, etc.) because the date input is being filled with an incorrect format - it expects "yyyy-MM-dd". Perhaps it's OK if we use the browser-accepted format for the input date, because the output date is human-friendly and that's the most important (from the user's point of view).

<img width="400" alt="date-calculator-screenshots" src="https://cloud.githubusercontent.com/assets/94173/17238436/9cef9956-5596-11e6-8887-00558705e7b9.png">

GuiltyDolphin commented:
@tagawa I'll just revert the changes to the date input - they can't complain if it is a text field.

Hmm... Locally the boxes wrap for me, but on Beta it's fine, weird... (I'm on FF).

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,37 @@
+    <p class="frm__desc pull--left tx--11">DURATION</p>
+    <span class="frm op--select">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="frm__select g fifty" style="height:2em">
```
use `!important` to override CSS -- inline styles shouldn't be necessary -- or give these a new class and create a more specific selector

GuiltyDolphin replied:
@moollaza See my comment above - I've tried using CSS with multiple classes + id + tag etc. (with `!important`) to no avail - unless I'm missing something obvious it looks like I can't override them?

moollaza replied:
`.zci--date_math .frm__select.g` seems to work?

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,37 @@
+<div class="gw .container" style="margin-bottom:1em">
+  <span class="input--date sixty">
+    <span class="frm g sixty">
```
You can remove this `<span>`

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,37 @@
+<div class="gw .container" style="margin-bottom:1em">
+  <span class="input--date sixty">
```
Whoops, meant you can remove this one, or at least the `sixty`. You can control the sizes of the other elements better to allow more space for the `<select>` at the end

moollaza replied:
I switched:
- the DATE input to `sixty`
- the `date--start` to `forty`
- `input--time` to `quarter`
- `input--op` to `quarter`

The HTML I ended up with looks like:

```html
<div class="gw container" style="margin-bottom:1em">
    <p class="frm__desc tx--11">
        <span class="frm g forty">DATE</span>
    </p>
    <span class="frm g forty">
        <input type="text" class="frm__input date--start" />
    </span>

    <p class="frm__desc tx--11">
        <span class="frm input--time input--numeric g quarter">TIME</span>
    </p>
    <span class="frm input--time input--numeric g quarter">
        <input type="text" maxlength="2" pattern="^(2[0-3]|[0-1]?[0-9])$" size="2" class="input--hour frm__input" />
        <span class="sep--time">:</span>
        <input type="text" maxlength="2" pattern="^([0-5][0-9]|[0-9])$" size="2" class="input--minute frm__input" />
        <span class="sep--time">:</span>
        <input type="text" maxlength="2" pattern="^([0-5][0-9]|[0-9])$" size="2" class="input--second frm__input" />
    </span>

    <p class="frm__desc tx--11">
        <span class="frm g ten">ACTION</span>
    </p>

    <p class="frm__desc pull--left tx--11">
        <span class="input--op g quarter" style="">DURATION</span>
    </p>
    <span class="input--op g quarter" style="">
        <span class="frm op--select">
            <input type="number" class= "frm__input input--op-amt forty" />
            <span class="frm__select sixty" style="height:2em">
                <select class="input--op-type" style="height:2em">
                    <option value="second">Seconds</option>
                    <option value="minute">Minutes</option>
                    <option value="hour">Hours</option>
                    <option value="day">Days</option>
                    <option value="week">Weeks</option>
                    <option value="month">Months</option>
                    <option value="year">Years</option>
                </select>
            </span>
        </span>
    </span>
</div>
```

GuiltyDolphin commented:
@moollaza Uhh, that appears to be missing the sign toggle? 

On Midori it takes it from:

![Uploading screenshot-2016-08-05-09:22:26_743_234.png…]()

To:

![Uploading screenshot-2016-08-05-09:22:46_744_347.png…]()

(I've played around with it a bit).

I'm not 100% sure what the issue is - on FF it looks screwed up locally, but on Beta it's fine. Could it just be something to do with slightly different styles on Beta and what DuckPAN receives?

moollaza commented:
> Uhh, that appears to be missing the sign toggle?

@GuiltyDolphin looks like your pictures didn't upload in that last comment?

> on FF it looks screwed up locally, but on Beta it's fine.

I'll take a look at FF...

> Could it just be something to do with slightly different styles on Beta and what DuckPAN receives?

Shouldn't be. If any, it would only be minor and temporary (once Beta is updated internally). AFAIK we haven't made any changes to CSS lately.

moollaza commented on share/goodie/date_math/content.handlebars:
```diff
@@ -0,0 +1,37 @@
+    <span class="frm op--select">
+      <input type="number" class="frm__input input--op-amt g forty"/>
+      <span class="frm__select g fifty" style="height:2em">
+        <select class="input--op-type" style="height:2em"/>
```
can use `.frm__select select.input--op-type` with `!important` here.

Our existing style has `!important` so we need to make our selector more specific first, then force it with `!important`

moollaza replied:
You'll want to also remove the padding on the `select` the `span` already adds some, forcing the `seconds` option to overlap with the dropdown

moollaza commented:
@GuiltyDolphin if you reduce the padding-left on the `.time--input .frm__input` then it should prevent the wrapping on FF. 

It may be wrapping due to some kind of calculation bug in FF... Not sure why the math isn't adding up there.

GuiltyDolphin commented:
@moollaza Everything just seems... Bigger, locally:

![screenshot-2016-08-06-18 35 19_857_220](https://cloud.githubusercontent.com/assets/8598426/17458284/a5ca92be-5c04-11e6-8708-af5b780a5c5e.png)

![screenshot-2016-08-06-18 35 33_870_274](https://cloud.githubusercontent.com/assets/8598426/17458286/ac94bf52-5c04-11e6-9535-22023c97f306.png)

(Beta/Local)

Changing anything just seems to screw it up more (locally); @tagawa was that screenshot with the wrapping on FF on beta or local?

mintsoft commented:
I'm going to close this for now until we decide what we want to do with it, feel free to re-open if you disagree
